clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -53,8 +53,10 @@
     private final Time time;
     private final long retryBackoffMs;
     private final long unsentExpiryMs;
-    // wakeup enabled flag need to be volatile since it is allowed to be accessed concurrently
-    volatile private boolean wakeupsEnabled = true;
+
+    // this count is only accessed from the consumer's main thread
+    private int wakeupDisabledCount = 0;
```

Comment:
There seems to be some theoretically possible edge cases where pollNoWakeup() can be called recursively. Admittedly they are far-fetched (e.g. having simultaneous autocommits and user async commits executing at the same time), but since the cost is small, it seemed worthwhile handling it.